plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -337,16 +365,39 @@ data Frame uni fun
 
 type Context uni fun = [Frame uni fun]
 
+{- Note [Catching exceptions in ST]
+This note represents MPJ's best understanding right now, might be wrong.
+
+We use a moderately evil trick to catch exceptions in ST. This uses the unsafe ST <-> IO conversion functions to go into IO,
+catch the exception, and then go back into ST.
+
+Why is this okay? Recall that IO ~= ST RealWorld, i.e. it is just ST with a special thread token. The unsafe conversion functions
+just coerve from one to the other. So the thread token remains the same, it's just that we'll potentially leak it from ST, and we don't
+get ordering guarantees with other IO actions.
+
+But in our case this is okay, because:
+
+1. We do not leak the original ST thread token, since we only pass it into IO and then immediately back again.
+2. We don't have ordering guarantees with other IO actions, but we don't care because we don't do any side effects, we only catch a single kind of exception.
+3. We *do* have ordering guarantees between the throws inside the ST action and the catch, since they are ultimately using the same thread token.
+-}
+
 runCekM
     :: forall a cost uni fun.
-       BuiltinsRuntime fun (CekValue uni fun)
+    (PrettyUni uni fun)
+    => BuiltinsRuntime fun (CekValue uni fun)
     -> ExBudgetMode cost uni fun
     -> Bool
     -> (forall s. CekM cost uni fun s a)
     -> (Either (CekEvaluationException uni fun) a, cost, [String])
 runCekM runtime (ExBudgetMode spender costInit) emitting a = runST $ do
     mayLogsRef <- if emitting then Just <$> newSTRef DList.empty else pure Nothing
-    (errOrRes, st') <- runStateT (runExceptT . runReaderT a $ CekEnv runtime mayLogsRef spender) costInit
+    let oldSt = costInit
+    -- See Note [Catching exceptions in ST]
+    maybeRes <- unsafeIOToST $ try @_ @(CekEvaluationException uni fun) $ unsafeSTToIO $ runStateT (runReaderT a $ CekEnv runtime mayLogsRef spender) costInit
```

Comment:
It's fine for now I guess, but we need to handle other exceptions too, not just `CekEvaluationException`. An accidental `error` might be triggered somewhere, for one example. But if we catch all of them, then we'll also catch async exceptions, which we shouldn't (and an exception thrown by `error` can be thrown asynchronously as well). Exception-handling is tricky...
I guess we need to talk to ledger people and ask them if they are fine with us catching all exceptions in the evaluator. But even if they are, we still need to distinguish between the normal `CekEvaluationException` and "something unexpected happened". So yeah, for now it's fine to only handle the former and bother about the latter later.

Reply:
Yeah, I think there are two problems here.

1. We want to throw and catch a specific, known set of errors, and we're using exceptions just so it's fast.
2. We want to ensure that we don't allow any unexpected exceptions to escape, so that they don't interfere with the operation of the ledger.

This is really only about 1. We could address 2 (which I think we have an issue for somewhere) by actually doing our own catching, but I think instead we should maybe just try hard to audit possible unexpected exceptions. As you point out, it's hard to not catch stuff you shouldn't catch, and the ledger is also vulnerable to their own "programmer error exceptions".

Reply:
> which I think we have an issue for somewhere

We do.

> instead we should maybe just try hard to audit possible unexpected exceptions.

This is quite unreliable I'd say. I'd rather catch exceptions (I'm talking about 2) and fail horribly on the ledger side if something does get caught. Either it's an async exception or an unexpected `error` and in both the cases you don't want to proceed as normal. Maybe up to reloading the node or something.

Reply:
It also depends on how this plutus machine is going to be integrated with the rest of the ledger code. Is the plutus machine going to be linked into existing ledger code or sandboxed somehow into a separate system process instance?

Reply:
It's going to be called as Haskell library code. So we do need to be careful about our exceptions escaping out.